### PR TITLE
[SPIR-V] Add SPV_INTEL_joint_matrix extension

### PR DESCRIPTION
**Pull request by MrSidims**
The spec is available here:
https://github.com/intel/llvm/pull/12497

The PR doesn't add OpCooperativeMatrixApplyFunctionINTEL instruction as it's still experimental and not properly tested E2E.

The PR also fixes few bugs in the related code:
1. CooperativeMatrixMulAddKHR optional operand must be literal, not a constant;
2. Fixed available capabilities table creation for a case, when a single extension adds few capabilities, that occupy not contiguous op codes.

### CODE REVIEW

**Comment by MrSidims** [comment]
@vmaksimo @VyacheslavLevytskyy please take a look

File: llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_joint_matrix/cooperative_matrix_bf16.ll [hunk 1]
```diff
@@ -0,0 +1,46 @@
+; RUN: llc -O0 -mtriple=spirv64-unknown-unknown --spirv-ext=+SPV_KHR_cooperative_matrix,+SPV_INTEL_joint_matrix %s -o - | FileCheck %s
```
Comment (VyacheslavLevytskyy) [inline llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_joint_matrix/cooperative_matrix_bf16.ll:1]: If it's something that spirv-val understands, we need to add also smth like

; RUN: %if spirv-tools %{ llc -O0 -mtriple=spirv64-unknown-unknown --spirv-ext=+SPV_KHR_cooperative_matrix,+SPV_INTEL_joint_matrix %s -o - -filetype=obj | spirv-val %}

in all new test cases.

Also -verify-machineinstrs if it doesn't fail.

File: llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_joint_matrix/cooperative_matrix_bf16.ll [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+; CHECK-DAG: Capability CooperativeMatrixKHR
+; CHECK-DAG: Extension "SPV_KHR_cooperative_matrix"
+; CHECK-DAG: Extension "SPV_INTEL_joint_matrix"
+; CHECK-DAG: CooperativeMatrixBFloat16ComponentTypeINTEL
+
+; CHECK: OpCooperativeMatrixMulAddKHR %[[#]] %[[#]] %[[#]] %[[#]] MatrixAAndBBFloat16ComponentsINTEL
+
+; ModuleID = 'test-matrix-opaque.bc'
```
Comment (VyacheslavLevytskyy) [inline llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_joint_matrix/cooperative_matrix_bf16.ll:10]: Let's please clean this and all following test cases of irrelevant details to keep it as clean and readable as we can -- unneeded comments, metadata records and references (if they are not require for the test) and source_filename, target datalayout and target triple.

File: llvm/lib/Target/SPIRV/SPIRVModuleAnalysis.cpp [hunk 3]
```diff
@@ -1437,6 +1437,114 @@ void addInstrRequirements(const MachineInstr &MI,
...
+  case SPIRV::OpCooperativeMatrixPrefetchINTEL: {
+    if (!ST.canUseExtension(SPIRV::Extension::SPV_KHR_cooperative_matrix))
+      report_fatal_error("Cooperative matrix instructions require the "
+                         "following SPIR-V extension: "
+                         "SPV_KHR_cooperative_matrix",
+                         false);
+    Reqs.addExtension(SPIRV::Extension::SPV_KHR_cooperative_matrix);
+    Reqs.addCapability(SPIRV::Capability::CooperativeMatrixKHR);
+
+    // Check Layout operand in case if it's not a standart one and add the
```
[the remark this answers is not in the document]
Comment (MrSidims) [reply llvm/lib/Target/SPIRV/SPIRVModuleAnalysis.cpp:1483]: I always have issues with this words as well as length <-> lenght, thanks!

File: llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_joint_matrix/cooperative_matrix_bf16.ll (hunk 1 above)
Comment (MrSidims) [reply llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_joint_matrix/cooperative_matrix_bf16.ll:1]: spirv-val won't understand it as the capabilities and the instructions are not yet propagated to SPIR-V Headers.

**Review by VyacheslavLevytskyy — APPROVED** [review]

**Review by michalpaszkowski — APPROVED** [review]

**Pull request merged**